cmd/dlv/cmds/commands.go

```diff
@@ -435,26 +446,47 @@ func dapCmd(cmd *cobra.Command, args []string) {
 			fmt.Fprintf(os.Stderr, "Warning: program flags ignored with dap; specify via launch/attach request instead\n")
 		}
 
-		listener, err := net.Listen("tcp", addr)
-		if err != nil {
-			fmt.Printf("couldn't start listener: %s\n", err)
-			return 1
-		}
+		var server *dap.Server
 		disconnectChan := make(chan struct{})
-		server := dap.NewServer(&service.Config{
-			Listener:       listener,
-			DisconnectChan: disconnectChan,
-			Debugger: debugger.Config{
-				Backend:              backend,
-				Foreground:           headless && tty == "",
-				DebugInfoDirectories: conf.DebugInfoDirectories,
-				CheckGoVersion:       checkGoVersion,
-				TTY:                  tty,
-			},
-			CheckLocalConnUser: checkLocalConnUser,
-		})
-		defer server.Stop()
 
+		if dapConnect == "" {
+			listener, err := net.Listen("tcp", addr)
+			if err != nil {
+				fmt.Printf("couldn't start listener: %s\n", err)
+				return 1
+			}
+			server = dap.NewServer(&service.Config{
+				Listener:       listener,
+				DisconnectChan: disconnectChan,
+				Debugger: debugger.Config{
+					Backend:              backend,
+					Foreground:           headless && tty == "",
+					DebugInfoDirectories: conf.DebugInfoDirectories,
+					CheckGoVersion:       checkGoVersion,
+					TTY:                  tty,
+				},
+				CheckLocalConnUser: checkLocalConnUser,
+			})
+		} else { // reverse mode
+			headless = true // TODO(github.com/go-delve/delve/issues/2552): consider the same for the normal mode.
+
+			conn, err := net.Dial("tcp", dapConnect)
```

Comment:
Sorry, I don't understand what you mean. There is nothing special here beyond the standard basic tcp connection and I don't see much of commonality beyond that.

preconnected Listener is used to arrange the dlv cli to connect to the server (possibly running in process). There, the client is still a client (the role in the rpc protocol) and the server is what uses service/debugger package. In this reverse mode, the entity that is dialing and plays the role of 'server' in the protocol, is what's using service/debugger package.

Reply:
In my understanding, the client is the entity that makes the debugging requests and the server is the entity that serves those requests. I don't think it matters which side connects to which side, client to server or server to client. Regardless of the specifics of that initial handshake, the rest of the traffic and meaning of the two sides don't change, do they? That's true for both rpc+terminal case and for dap+special-client case.
 
In this new mode you are defining, you have a predefined client that you connect to before you call NewServer.
In case of the rpc server + terminal client, we also have a client connection that is [predefined and pre-connected](https://github.com/go-delve/delve/blob/38aaf274a92fc620119bdf257500ee3b3f677774/cmd/dlv/cmds/commands.go#L820) before newServer is called. Yet newServer still takes a listener, but that listener is fake. The listener doesn't actually listen as it is already pre-connected to the connection point that was already set up. On the very first call to Accept it will return that pre-existing client connection. And because of that abstraction, newServer and Run can be re-used as-is without any special logic or comments inside of the server itself. 

I propose that we do the same here to move connection details outside of the server, keeping it general and consistent with the rpc server, which we ultimately want to merge with under a single command.

`listenerpipe.go` has `ListenerPipe` + generic `preconnectedListener`, which you can reuse more or less ass-is. You can define something like this:

```
func ListenerFromTCPClient(clientAddr string) net.Listener { 
         conn, err := net.Dial("tcp", clientAddr)
	return &preconnectedListener{conn: conn, closech: make(chan struct{})}
}
```
Then pass the result to `NewServer` and not need to modify the dap server code.

Reply:
I think the proposed approach overcomplicates the code without clear use cases. When the use case that benefits from it is obvious, I am ok to rework this. Until then, I prefer simplicity and less channels and locks.